### PR TITLE
Add tasklist sample.

### PR DESCRIPTION
**Pull request by jeffmendoza**
also updated gems.

### CODE REVIEW

File: datastore/spec/tasks_spec.rb [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    id = task.key.id
+    mark_done project_id: @project_id,
+              task_id: id
+    task = @datastore.find "Task", id
+    expect(task["done"]).to be(true)
+  end
+
+  it "list_tasks" do
+    task1 = @datastore.entity "Task" do |t|
+      t["description"] = "Test 1."
```
Comment (frankyn) [inline datastore/spec/tasks_spec.rb]: Align your "=" here.

File: datastore/spec/tasks_spec.rb [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+      t["created"] = Time.now
+      t["done"] = false
+      t.exclude_from_indexes! "description", true
+    end
+    task2 = @datastore.entity "Task" do |t|
+      t["description"] = "Test 2."
```
Comment (frankyn) [inline datastore/spec/tasks_spec.rb]: Align "=" here as well

File: datastore/spec/tasks_spec.rb [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+      t.exclude_from_indexes! "description", true
+    end
+    @datastore.save task1, task2
+
+    expect { list_tasks project_id: @project_id }.to output(/Test 1/).to_stdout
+  end
+
+  it "deletes tasks" do
+    task = @datastore.entity "Task" do |t|
+      t["description"] = "Test 1."
```
Comment (frankyn) [inline datastore/spec/tasks_spec.rb]: Align "=" here

File: datastore/tasks.rb [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+  gcloud.datastore
+end
+# [END build_service]
+
+# [START add_entity]
+def new_task project_id:, description:
+  datastore = create_client project_id: project_id
+
+  task = datastore.entity "Task" do |t|
+    t["description"] = description
```
Comment (frankyn) [inline datastore/tasks.rb]: Align "="

File: datastore/tasks.rb [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+  task["done"] = true
+  datastore.save task
+end
+# [END update_entity]
+
+# [START retrieve_entities]
+def list_tasks project_id:
+  datastore = create_client project_id: project_id
+
+  query = datastore.query("Task").
```
Comment (frankyn) [inline datastore/tasks.rb]: This should be able to fit in one line.

File: datastore/spec/tasks_spec.rb [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+# limitations under the License.
+
+require_relative "../tasks"
+require "rspec"
+
+describe "Datastore task list" do
+
+  before :all do
+    @project_id = ENV["GOOGLE_CLOUD_PROJECT"]
+    @datastore = Google::Cloud::Datastore.new project: @project_id
```
Comment (frankyn) [inline datastore/spec/tasks_spec.rb]: Align "="

File: datastore/spec/tasks_spec.rb [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+  it "creates client" do
+    datastore = create_client project_id: @project_id
+  end
+
+  it "creates a task" do
+    desc = "Test description."
+    allow($stdout).to receive(:puts)
+    id = new_task project_id: @project_id,
+                    description: desc
```
Comment (frankyn) [inline datastore/spec/tasks_spec.rb]: Align parameters and values

File: datastore/tasks.rb [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+  project_id = ENV["GOOGLE_CLOUD_PROJECT"]
+  case ARGV.shift
+  when "new"
+    new_task project_id: project_id,
+             description: ARGV.shift
+  when "done"
+    mark_done project_id:  project_id,
+              task_id: ARGV.shift
+  when "list"
+    list_tasks project_id:  project_id
```
Comment (frankyn) [inline datastore/tasks.rb]: Delete double space

File: datastore/tasks.rb [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+  when "delete"
+    delete_task project_id:  project_id,
```
Comment (frankyn) [inline datastore/tasks.rb]: Align parameter value

File: datastore/tasks.rb (hunk 4 above)
Comment (frankyn) [inline datastore/tasks.rb]: align "="

File: datastore/tasks.rb [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+  puts task.key.id
+  task.key.id
+end
+# [END add_entity]
+
+# [START update_entity]
+def mark_done project_id:, task_id:
+  datastore = create_client project_id: project_id
+
+  task = datastore.find "Task", task_id.to_i
```
Comment (frankyn) [inline datastore/tasks.rb]: align "="

File: datastore/spec/tasks_spec.rb (hunk 1 above)
Comment (frankyn) [inline datastore/spec/tasks_spec.rb]: align "="

File: datastore/spec/tasks_spec.rb (hunk 2 above)
Comment (frankyn) [inline datastore/spec/tasks_spec.rb]: align "="

File: datastore/spec/tasks_spec.rb (hunk 3 above)
Comment (frankyn) [inline datastore/spec/tasks_spec.rb]: align "="

File: datastore/spec/tasks_spec.rb [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+  it "deletes tasks" do
+    task = @datastore.entity "Task" do |t|
+      t["description"] = "Test 1."
+      t["created"] = Time.now
+      t["done"] = false
+      t.exclude_from_indexes! "description", true
+    end
+    @datastore.save task
+    id = task.key.id
+    delete_task project_id: @project_id,
```
Comment (frankyn) [inline datastore/spec/tasks_spec.rb]: align parameter values

File: datastore/tasks.rb [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+  datastore.delete task
+end
+# [END delete_entity]
+
+if __FILE__ == $0
+  project_id = ENV["GOOGLE_CLOUD_PROJECT"]
+  case ARGV.shift
+  when "new"
+    new_task project_id: project_id,
```
Comment (frankyn) [inline datastore/tasks.rb]: align parameter value

File: datastore/tasks.rb [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+             description: ARGV.shift
+  when "done"
+    mark_done project_id:  project_id,
```
Comment (frankyn) [inline datastore/tasks.rb]: Align parameter value

File: datastore/tasks.rb (hunk 5 above)
Comment (jeffmendoza) [inline datastore/tasks.rb]: For the datastore queries, it seems like a pattern to put every .(something) on a new line
https://cloud.google.com/datastore/docs/concepts/queries#datastore-basic-query-ruby
there are more examples with 1 or 2 where, order, etc. on new lines

**Comment by jeffmendoza** [comment]
Updated, thanks!

**Comment by remi** [comment]
Apologies for the lack of clarity around our style for code snippets.

The samples in this repository use an inconsistent syntax and template ~ this is something we're looking into fixing soon!

Everything looks great!  I would like to change one thing about the sample, which is how the client library for datastore is retrieved in each method.

Rather than this:

```ruby
def mark_done project_id:, task_id:
  datastore = create_client project_id: project_id

  task         = datastore.find "Task", task_id.to_i
  task["done"] = true
  datastore.save task
end
```

I would like to update it to this:

```ruby
def mark_done task_id
  task = @datastore.find "Task", task_id

  task["done"] = true

  datastore.save task
end
```

Primarily, I would like to set a shared variable so that the client library is available to all of the methods.

See the [Tutorial application snippets](https://github.com/GoogleCloudPlatform/ruby-docs-samples/blob/tswast-styleguide/CONTRIBUTING.md#tutorial-application-snippets) documentation in CONTRIBUTING.md (*currently proposed by pull #44 to help add clarity around this*)

Sidenotes:

 - `mark_done task_id` is clear, so let's not use a named parameter
 - the alignment feels a little awkward - let's just split these out onto 3 separate lines to clearly show (1) fetch (2) update object (3) commit to datastore
 - instead of `task_id.to_i`, in an application you would presumably only call `mark_done` with a valid parameter, so let's remove the casting

Apologies again for the confusion around syntax!  We're working on this.  The logic looks great!  😄 

Let me know if the updates are something you would like assistance with

**Comment by jeffmendoza** [comment]
@remi How does it look now? closer? =)

**Comment by remi** [comment]
LGTM! 👍 

#116 updated the google-cloud-datastore dependency ~ this needs to be updated before it can be merged

**Comment by remi** [comment]
Merged master (needed to resolve conflict with Gemfile.lock) ~ ready to go!

**Pull request merged**